src/main/java/org/prebid/server/analytics/reporter/greenbids/GreenbidsAnalyticsReporter.java

```diff
@@ -0,0 +1,355 @@
+                .map(Banner::getFormat)
+                .map(formats -> formats.stream().map(format -> List.of(format.getW(), format.getH())))
+                .orElse(null).collect(Collectors.toList());
+
+        final ExtBanner extBanner = ExtBanner.builder()
+                .sizes(bannerWidthHeight)
+                .pos(Optional.ofNullable(banner).map(Banner::getPos).orElse(null))
+                .name(Optional.ofNullable(banner).map(Banner::getId).orElse(null))
+                .build();
+
+        final MediaTypes mediaTypes = MediaTypes.builder()
+                .banner(extBanner)
+                .video(video)
+                .nativeObject(nativeObject)
+                .build();
+
+        final List<GreenbidsBids> bids = extractBidders(imp, seatsWithBids, seatsWithNonBids);
+
+        final String adUnitCode = getAdUnitCode(imp);
+
+        return AdUnit.builder()
+                .code(adUnitCode)
+                .mediaTypes(mediaTypes)
+                .bids(bids)
+                .build();
+    }
+
+    private List<GreenbidsBids> extractBidders(
+            Imp imp, Map<String, Bid> seatsWithBids, Map<String, NonBid> seatsWithNonBids) {
+        final List<GreenbidsBids> bidders = new ArrayList<>();
+
+        final Map<String, Bid> seatsWithBidsForImp = seatsWithBids.entrySet().stream()
+                .filter(entry -> entry.getValue().getImpid().equals(imp.getId()))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        final Map<String, NonBid> seatsWithNonBidsForImp = seatsWithNonBids.entrySet().stream()
+                .filter(entry -> entry.getValue().getImpId().equals(imp.getId()))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        seatsWithBidsForImp.forEach((seat, bid) -> {
+            final GreenbidsBids bidder = GreenbidsBids.ofBid(seat, bid);
+            bidders.add(bidder);
+        });
+
+        seatsWithNonBidsForImp.forEach((seat, nonBid) -> {
+            final GreenbidsBids bidder = GreenbidsBids.ofNonBid(seat, nonBid);
+            bidders.add(bidder);
+        });
+
+        return bidders;
+    }
+
+    private String getAdUnitCode(Imp imp) {
+        final ObjectNode impExt = imp.getExt();
+
+        return Optional.ofNullable(getGpid(impExt))
+                .or(() -> Optional.ofNullable(storedRequestId(impExt)))
+                .orElse(imp.getId());
+    }
+
+    private static String getGpid(ObjectNode impExt) {
+        return Optional.ofNullable(impExt)
+                .map(ext -> ext.get("gpid"))
+                .map(JsonNode::asText)
+                .orElse(null);
+    }
+
+    private String storedRequestId(ObjectNode impExt) {
+        return Optional.ofNullable(impExt)
+                .map(ext -> ext.get("prebid"))
+                .map(this::extImpPrebid)
+                .map(ExtImpPrebid::getStoredrequest)
+                .map(ExtStoredRequest::getId)
+                .orElse(null);
+    }
```

Comment:
private String getAdUnitCode(Imp imp) {
        final ObjectNode impExt = imp.getExt();

        return getGpid(impExt)
                .or(() -> getStoredRequestId(impExt))
                .orElse(imp.getId());
    }

    private static Optional<String> getGpid(ObjectNode impExt) {
        return Optional.ofNullable(impExt)
                .map(ext -> ext.get("gpid"))
                .map(JsonNode::asText);
    }

    private Optional<String> getStoredRequestId(ObjectNode impExt) {
        return Optional.ofNullable(impExt)
                .map(ext -> ext.get("prebid"))
                .map(this::extImpPrebid)
                .map(ExtImpPrebid::getStoredrequest)
                .map(ExtStoredRequest::getId);
    }